### PR TITLE
QFJ-964: Fixing a bug where sections inside SessionSettings is not th…

### PR DESCRIPTION
**Pull request by alexwibowo**
This is to fix a bug as defined in https://www.quickfixj.org/jira/browse/QFJ-964

### CODE REVIEW

File: quickfixj-core/src/test/java/quickfix/SessionSettingsTest.java [hunk 1]
```diff
@@ -390,4 +389,120 @@ public void testParseSettingReconnectInterval() {
...
+            new Thread(() -> {
+                final Map<Object, Object> expectedClientPricingSettings = new HashMap<>();
+                expectedClientPricingSettings.putAll(defaultSettings);
+                expectedClientPricingSettings.putAll(pricingSection);
+
+                final Map<Object, Object> expectedClientTradingSettings = new HashMap<>();
+                expectedClientTradingSettings.putAll(defaultSettings);
+                expectedClientTradingSettings.putAll(tradingSection);
+
+                int randomSleep = random.nextInt(50);
```
[the remark this answers is not in the document]
Comment (chrjohn) [reply quickfixj-core/src/test/java/quickfix/SessionSettingsTest.java]: When I said "failed" I meant of course before changing the `sections` to `ConcurrentHashMap`.

File: quickfixj-core/src/test/java/quickfix/SessionSettingsTest.java [hunk 2]
```diff
@@ -390,4 +389,120 @@ public void testParseSettingReconnectInterval() {
...
+                    final SessionID clientTradingSessionID = new SessionID(clientTradingSessionIDString);
+                    sessionSettings.set(clientTradingSessionID, new Dictionary(clientTradingSessionIDString, expectedClientTradingSettings ));
+
+                    // sleep at the end, before we verify the outcome
+                    Thread.sleep(randomSleep);
+
+                    assertEquals("Default settings must be correct", defaultSettings, sessionSettings.get().toMap());
+                    assertEquals("Client pricing settings must be correct", expectedClientPricingSettings, sessionSettings.get(clientPricingSessionID).toMap());
+                    assertEquals("Client trading settings must be correct",expectedClientTradingSettings, sessionSettings.get(clientTradingSessionID).toMap());
+                } catch (Exception exception) {
```
[the remark this answers is not in the document]
Comment (chrjohn) [reply quickfixj-core/src/test/java/quickfix/SessionSettingsTest.java]: Example:

```
Running quickfix.SessionSettingsTest
Exception in thread "Thread-11" Exception in thread "Thread-318" junit.framework.AssertionFailedError: Default settings must be correct expected:<{UseJmx=Y, BeginString=FIX.4.2, EndTime=17:00:00, ConnectionType=acceptor, FileStorePath=/home/wibowoa/var/lib/myApp, TargetCompID=*, StartTime=7:00:00, StartDay=Sunday, UseDataDictionary=Y, FileIncludeTimeStampForMessages=Y, CheckLatency=Y, TimeZone=UTC, NonStopSession=N, EndDay=Friday, FileIncludeMilliseconds=Y, AcceptorTemplate=Y, HeartBtInt=30, ThreadModel=ThreadPerSession, FileLogPath=logs/fixlog}> but was:<{}>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.TestCase.assertEquals(TestCase.java:244)
	at quickfix.SessionSettingsTest.lambda$testConcurrentAccess$0(SessionSettingsTest.java:435)
	at java.lang.Thread.run(Thread.java:745)
junit.framework.AssertionFailedError: Default settings must be correct expected:<{UseJmx=Y, BeginString=FIX.4.2, EndTime=17:00:00, ConnectionType=acceptor, FileStorePath=/home/wibowoa/var/lib/myApp, TargetCompID=*, StartTime=7:00:00, StartDay=Sunday, UseDataDictionary=Y, FileIncludeTimeStampForMessages=Y, CheckLatency=Y, TimeZone=UTC, NonStopSession=N, EndDay=Friday, FileIncludeMilliseconds=Y, AcceptorTemplate=Y, HeartBtInt=30, ThreadModel=ThreadPerSession, FileLogPath=logs/fixlog}> but was:<{}>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.TestCase.assertEquals(TestCase.java:244)
	at quickfix.SessionSettingsTest.lambda$testConcurrentAccess$0(SessionSettingsTest.java:435)
	at java.lang.Thread.run(Thread.java:745)
Tests run: 20, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.639 s - in quickfix.SessionSettingsTest
```

The errors are logged but the test is not failing.

File: quickfixj-core/src/main/java/quickfix/SessionSettings.java [hunk 3]
```diff
@@ -365,7 +366,7 @@ public void setBool(SessionID sessionID, String key, boolean value) {
         getOrCreateSessionProperties(sessionID).setProperty(key, BooleanConverter.convert(value));
     }
 
-    private final HashMap<SessionID, Properties> sections = new HashMap<>();
+    private final ConcurrentHashMap<SessionID, Properties> sections = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (alexwibowo) [reply quickfixj-core/src/main/java/quickfix/SessionSettings.java]: Done

File: quickfixj-core/src/test/java/quickfix/SessionSettingsTest.java [hunk 4]
```diff
@@ -390,4 +389,120 @@ public void testParseSettingReconnectInterval() {
...
+        final int numClients = 500;
+        final CountDownLatch startLatch = new CountDownLatch(1);
+        final CountDownLatch countDownLatch = new CountDownLatch(numClients);
+
+        final AtomicBoolean testHasPassed = new AtomicBoolean(true);
+        final Random random = new Random();
+        for (int i = 0; i < numClients; i++) {
+            final String clientPricingSessionIDString = "FIX.4.2:FOOBAR_PRICING->CLIENT" + i;
+            final String clientTradingSessionIDString = "FIX.4.2:FOOBAR_TRADING->CLIENT" + i;
+            new Thread(() -> {
```
[the remark this answers is not in the document]
Comment (alexwibowo) [reply quickfixj-core/src/test/java/quickfix/SessionSettingsTest.java]: I tried to use PausableThreadPoolExecutor, following example on SessionDisconnectConcurrentlyTest. But I must not have used it correctly. So for now i just set the thread to be daemon thread.

**Pull request merged**